### ISSUE TITLE
Panic observed in kube.(*logsReader).GetLogsByJobAndContainerName

### ISSUE THREAD

**Issue by mag009** [aquasecurity/starboard#627]
**What steps did you take and what happened:**

I tried with one namespace and it worked the operator does not crash. But for one particular namespace the operator crash
it seem to crash on a particular cronjob as soon as i delete the cron the operator work.

**What did you expect to happen:**
Jobs to trigger vulnerabilities on all namespaces.

- Starboard version : 0.10.3
- Kubernetes version : v1.20.4-eks-6b7464
- EKS

```{"level":"info","ts":1624461870.6293688,"logger":"controller-runtime.manager.controller.daemonset","msg":"Starting workers","reconciler group":"apps","reconciler kind":"DaemonSet","worker count":1}
E0623 15:24:30.674204 1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 552 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x15e2ba0, 0x22e29c0)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:48 +0x86
panic(0x15e2ba0, 0x22e29c0)
/opt/hostedtoolcache/go/1.16.4/x64/src/runtime/panic.go:965 +0x1b9
github.com/aquasecurity/starboard/pkg/kube.(*logsReader).GetLogsByJobAndContainerName(0xc0000a5a30, 0x1984298, 0xc0010fa630, 0xc000fd8000, 0xc000c8cd68, 0x17, 0x17, 0xc0007d85b8, 0x14, 0xc00004c8d0)
/home/runner/work/starboard/starboard/pkg/kube/logs.go:35 +0x1e5
github.com/aquasecurity/starboard/pkg/operator/controller.(*VulnerabilityReportReconciler).processCompleteScanJob(0xc0001907e0, 0x1984298, 0xc0010fa630, 0xc000fd8000, 0x9, 0xc0007cebd0)
/home/runner/work/starboard/starboard/pkg/operator/controller/vulnerabilityreport.go:375 +0xb28
github.com/aquasecurity/starboard/pkg/operator/controller.(*VulnerabilityReportReconciler).reconcileJobs.func1(0x1984298, 0xc0010fa630, 0xc00004c850, 0x9, 0xc0007cebd0, 0x22, 0x16ab240, 0xc0010fa630, 0x15e11c0, 0x2331ca8)
/home/runner/work/starboard/starboard/pkg/operator/controller/vulnerabilityreport.go:323 +0x31b
sigs.k8s.io/controller-runtime/pkg/reconcile.Func.Reconcile(0xc00042a020, 0x1984298, 0xc0010fa630, 0xc00004c850, 0x9, 0xc0007cebd0, 0x22, 0xc0010fa630, 0xc000030000, 0x166c580, ...)
/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/reconcile/reconcile.go:102 +0x6f
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00017d680, 0x19841f0, 0xc000295b00, 0x1635620, 0xc000cdbba0)
/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:263 +0x30d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00017d680, 0x19841f0, 0xc000295b00, 0xc000363600)
/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:235 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1.1(0x19841f0, 0xc000295b00)
/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:198 +0x4a
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0x37
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc000363750)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:155 +0x5f
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc002235f50, 0x1955bc0, 0xc0010fa570, 0xc000295b01, 0xc0000489c0)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:156 +0x9b
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000363750, 0x3b9aca00, 0x0, 0x1270101, 0xc0000489c0)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:133 +0x98
k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x19841f0, 0xc000295b00, 0xc00042b7b0, 0x3b9aca00, 0x0, 0x1)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:185 +0xa6
k8s.io/apimachinery/pkg/util/wait.UntilWithContext(0x19841f0, 0xc000295b00, 0xc00042b7b0, 0x3b9aca00)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/wait/wait.go:99 +0x57
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
/home/runner/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.7.2/pkg/internal/controller/controller.go:195 +0x497
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x13a0cc5]```


I've tested with version 0.10.1 and it is working seem to have been introduced in version 0.10.2+

**Comment by danielpacak** [comment]
It seems that this issue is not related to target namespace or a difference between v0.10.1 and v0.10.3 but the way the operator reads logs. The code panics when we are trying to retrieve logs when scan job is completed by scan job name and container name.

```
E0623 15:24:30.674204 1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 552 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x15e2ba0, 0x22e29c0)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:74 +0xa6
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
/home/runner/go/pkg/mod/k8s.io/apimachinery@v0.19.2/pkg/util/runtime/runtime.go:48 +0x86
panic(0x15e2ba0, 0x22e29c0)
/opt/hostedtoolcache/go/1.16.4/x64/src/runtime/panic.go:965 +0x1b9
github.com/aquasecurity/starboard/pkg/kube.(*logsReader).GetLogsByJobAndContainerName(0xc0000a5a30, 0x1984298, 0xc0010fa630, 0xc000fd8000, 0xc000c8cd68, 0x17, 0x17, 0xc0007d85b8, 0x14, 0xc00004c8d0)
/home/runner/work/starboard/starboard/pkg/kube/logs.go:35 +0x1e5
github.com/aquasecurity/starboard/pkg/operator/controller.(*VulnerabilityReportReconciler).processCompleteScanJob(0xc0001907e0, 0x1984298, 0xc0010fa630, 0xc000fd8000, 0x9, 0xc0007cebd0)
/home/runner/work/starboard/starboard/pkg/operator/controller/vulnerabilityreport.go:375 +0xb28
```

We do handle errors in `GetLogsByJobAndContainerName()` but we do not recover from panic that might originate from the k8s.io/apimachinery@v0.19.2 module.

The code that panic instead of returning an error is the implementation of the `kube.LogsReader` interface. We need to figure the reason for panic.

```go
return r.clientset.CoreV1().Pods(pod.Namespace).
	GetLogs(pod.Name, &corev1.PodLogOptions{
		Follow:    true,
		Container: containerName,
	}).Stream(ctx)
```

**Comment by travisghansen** [comment]
Pretty sure I've got the same thing going on. If you need further details let me know if I can provide extra info. I've been in a crashloop backoff for a good month and a half now :D

**Comment by danielpacak** [comment]
> Pretty sure I've got the same thing going on. If you need further details let me know if I can provide extra info. I've been in a crashloop backoff for a good month and a half now :D

:wave: @travisghansen Do you see by any chance what kind of K8s workload is causing this problem? Is it Job or CronJob or all types such as Pod, ReplicaSet, DaemonSeet, etc.?

**Comment by travisghansen** [comment]
@danielpacak I'm not sure how I would discover that? Give me some tips and I'll see what I can find. In basically every cluster I've installed the operator (~10) in it's sitting a crashloop backoff with this error. This probably isn't helpful but 2 of the clusters I looked at showed statefulset as the most recent log entry and the other showed cronjob (this is the 'starting workers' message).

**Comment by danielpacak** [comment]
@matzegebbe Thank you for the feedback. I think we should add more logs around the code that's accessing pod logs. Also we could try recovering from panic and ignore it gracefully. I'm afraid that the panic comes from the controller-runtime or client-go package, not the Starboard code.

**Comment by danielpacak** [comment]
Fixed by #1700a32 and client-go and controller-runtime bump ups. Please reopen if the problem persists.

**Comment by danielpacak** [comment]
> Is it in a released version to test with?

It will be available in v0.12.0 that we are going to release soon (days).
